cpp/src/arrow/compute/exec/query_context.cc

```diff
@@ -0,0 +1,96 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "arrow/compute/exec/query_context.h"
+#include "arrow/util/cpu_info.h"
+#include "arrow/util/io_util.h"
+
+namespace arrow {
+using internal::CpuInfo;
+namespace compute {
+QueryOptions::QueryOptions()
+    : max_memory_bytes(
+          static_cast<size_t>(0.75f * ::arrow::internal::GetTotalMemoryBytes())),
+      use_legacy_batching(false) {}
+
+QueryContext::QueryContext(QueryOptions opts, ExecContext exec_context)
+    : options_(opts),
+      exec_context_(exec_context),
+      io_context_(exec_context_.memory_pool()) {}
```

Comment:
This means we are always using the default I/O thread pool...which is probably ok for now.  I'm not sure who is supposed to be providing the I/O context anyways (e.g. filesystem?)

Reply:
Yeah I figure this won't be hard to change later once we figure out what we want to be configurable.